File: packages/python/src/alumnium/clients/http_client.py
from requests import delete, get, post

from ..server.models import Model
from ..tools.base_tool import BaseTool
from ..tools.tool_to_schema_converter import convert_tools_to_schemas
from .typecasting import Data, loosely_typecast


class HttpClient:
    def __init__(
        self,
        base_url: str,
        model: Model,
        platform: str,
        tools: dict[str, type[BaseTool]],
        planner: bool = True,
    ):
        self.base_url = base_url.rstrip("/")
        self.session_id = None

        tool_schemas = convert_tools_to_schemas(tools)

        response = post(
            f"{self.base_url}/v1/sessions",
            json={
                "provider": model.provider.value,
                "name": model.name,
                "tools": tool_schemas,
                "platform": platform,
                "planner": planner,
            },
            timeout=30,
        )
        response.raise_for_status()
        self.session_id = response.json()["session_id"]

    def quit(self):
        if self.session_id:
            response = delete(
                f"{self.base_url}/v1/sessions/{self.session_id}",
                timeout=30,
            )
            response.raise_for_status()
            self.session_id = None

    def plan_actions(self, goal: str, accessibility_tree: str) -> tuple[str, list[str]]:
        """
        Plan actions to achieve a goal.
        Returns:
            A tuple of (explanation, steps).
        """
        response = post(
            f"{self.base_url}/v1/sessions/{self.session_id}/plans",
            json={"goal": goal, "accessibility_tree": accessibility_tree},
            timeout=120,
        )
        response.raise_for_status()
        response_data = response.json()
        return (response_data["explanation"], response_data["steps"])

    def add_example(self, goal: str, actions: list[str]):
        response = post(
            f"{self.base_url}/v1/sessions/{self.session_id}/examples",
            json={"goal": goal, "actions": actions},
            timeout=30,
        )
        response.raise_for_status()
        return response.json()

    def clear_examples(self):
        response = delete(
            f"{self.base_url}/v1/sessions/{self.session_id}/examples",
            timeout=30,
        )
        response.raise_for_status()

    def execute_action(self, goal: str, step: str, accessibility_tree: str) -> tuple[str, list[dict]]:
        response = post(
            f"{self.base_url}/v1/sessions/{self.session_id}/steps",
            json={"goal": goal, "step": step, "accessibility_tree": accessibility_tree},
            timeout=120,
        )
        response.raise_for_status()
        data = response.json()
        return data["explanation"], data["actions"]

    def retrieve(
        self,
        statement: str,
        accessibility_tree: str,
        title: str,
        url: str,
        screenshot: str | None,
    ) -> tuple[str, Data]:
        response = post(
            f"{self.base_url}/v1/sessions/{self.session_id}/statements",
            json={
                "statement": statement,
                "accessibility_tree": accessibility_tree,
                "title": title,
                "url": url,
                "screenshot": screenshot if screenshot else None,
            },
            timeout=120,
        )
        response.raise_for_status()
        data = response.json()
        return data["explanation"], loosely_typecast(data["result"])

    def find_area(self, description: str, accessibility_tree: str):
        response = post(
            f"{self.base_url}/v1/sessions/{self.session_id}/areas",
            json={"description": description, "accessibility_tree": accessibility_tree},
            timeout=60,
        )
        response.raise_for_status()
        data = response.json()
        return {"id": data["id"], "explanation": data["explanation"]}

    def find_element(self, description: str, accessibility_tree: str) -> dict:
        response = post(
            f"{self.base_url}/v1/sessions/{self.session_id}/elements",
            json={"description": description, "accessibility_tree": accessibility_tree},
            timeout=60,
        )
        response.raise_for_status()
        return response.json()["elements"][0]

    def analyze_changes(
        self,
        before_accessibility_tree: str,
        before_url: str,
        after_accessibility_tree: str,
        after_url: str,
    ) -> str:
        response = post(
            f"{self.base_url}/v1/sessions/{self.session_id}/changes",
            json={
                "before": {
                    "accessibility_tree": before_accessibility_tree,
                    "url": before_url,
                },
                "after": {
                    "accessibility_tree": after_accessibility_tree,
                    "url": after_url,
                },
            },
            timeout=120,
        )
        response.raise_for_status()
        return response.json()["result"]

    def save_cache(self):
        response = post(
            f"{self.base_url}/v1/sessions/{self.session_id}/caches",
            timeout=30,
        )
        response.raise_for_status()

    def discard_cache(self):
        response = delete(
            f"{self.base_url}/v1/sessions/{self.session_id}/caches",
            timeout=30,
        )
        response.raise_for_status()

    @property
    def stats(self):
        response = get(
            f"{self.base_url}/v1/sessions/{self.session_id}/stats",
            timeout=30,
        )
        response.raise_for_status()
        return response.json()

File: packages/typescript/src/clients/HttpClient.ts
import { Model } from "../Model.js";
import { ToolClass } from "../tools/BaseTool.js";
import { convertToolsToSchemas } from "../tools/toolToSchemaConverter.js";
import { getLogger } from "../utils/logger.js";
import {
  AddExampleRequest,
  AreaRequest,
  AreaResponse,
  FindRequest,
  FindResponse,
  PlanRequest,
  PlanResponse,
  SessionRequest,
  SessionResponse,
  StatementRequest,
  StatementResponse,
  StatsResponse,
  StepRequest,
  StepResponse,
} from "./ApiModels.js";
import { Data, looselyTypecast } from "./typecasting.js";

const logger = getLogger(["HttpClient"]);

export class HttpClient {
  private baseUrl: string;
  private sessionId: string | null = null;
  private sessionPromise: Promise<void> | null = null;
  private timeout: number = 300_000; // 5 minutes

  constructor(
    baseUrl: string,
    private model: Model,
    private platform: string,
    private tools: Record<string, ToolClass>,
    private planner: boolean = true
  ) {
    this.baseUrl = baseUrl.replace(/\/$/, "");
  }

  private async fetchWithTimeout(
    url: string,
    options: RequestInit = {}
  ): Promise<Response> {
    const response = await fetch(url, {
      ...options,
      signal: AbortSignal.timeout(this.timeout),
    });

    if (!response.ok) {
      throw new Error(`HTTP error! status: ${response.status}`);
    }

    return response;
  }

  private async ensureSession(): Promise<void> {
    if (this.sessionId) {
      return;
    }

    if (!this.sessionPromise) {
      this.sessionPromise = (async () => {
        const toolSchemas = convertToolsToSchemas(this.tools);
        const requestBody: SessionRequest = {
          provider: this.model.provider,
          name: this.model.name,
          platform: this.platform as SessionRequest["platform"],
          tools: toolSchemas,
          planner: this.planner,
        };
        const response = await this.fetchWithTimeout(
          `${this.baseUrl}/v1/sessions`,
          {
            method: "POST",
            headers: {
              "Content-Type": "application/json",
            },
            body: JSON.stringify(requestBody),
          }
        );

        const data = (await response.json()) as SessionResponse;
        this.sessionId = data.session_id;
        logger.debug(`Session initialized with ID: ${this.sessionId}`);
      })();
    }

    await this.sessionPromise;
  }

  async quit(): Promise<void> {
    if (this.sessionId) {
      await this.fetchWithTimeout(
        `${this.baseUrl}/v1/sessions/${this.sessionId}`,
        {
          method: "DELETE",
        }
      );
      this.sessionId = null;
    }
  }

  async planActions(
    goal: string,
    accessibilityTree: string
  ): Promise<{ explanation: string; steps: string[] }> {
    await this.ensureSession();
    const requestBody: PlanRequest = {
      goal,
      accessibility_tree: accessibilityTree,
    };
    const response = await this.fetchWithTimeout(
      `${this.baseUrl}/v1/sessions/${this.sessionId}/plans`,
      {
        method: "POST",
        headers: {
          "Content-Type": "application/json",
        },
        body: JSON.stringify(requestBody),
      }
    );

    const responseData = (await response.json()) as PlanResponse;
    return { explanation: responseData.explanation, steps: responseData.steps };
  }

  async addExample(goal: string, actions: string[]): Promise<void> {
    await this.ensureSession();
    const requestBody: AddExampleRequest = {
      goal,
      actions,
    };
    await this.fetchWithTimeout(
      `${this.baseUrl}/v1/sessions/${this.sessionId}/examples`,
      {
        method: "POST",
        headers: {
          "Content-Type": "application/json",
        },
        body: JSON.stringify(requestBody),
      }
    );
  }

  async clearExamples(): Promise<void> {
    await this.ensureSession();
    await this.fetchWithTimeout(
      `${this.baseUrl}/v1/sessions/${this.sessionId}/examples`,
      {
        method: "DELETE",
      }
    );
  }

  async executeAction(
    goal: string,
    step: string,
    accessibilityTree: string
  ): Promise<{ explanation: string; actions: StepResponse["actions"] }> {
    await this.ensureSession();
    const requestBody: StepRequest = {
      goal,
      step,
      accessibility_tree: accessibilityTree,
    };
    const response = await this.fetchWithTimeout(
      `${this.baseUrl}/v1/sessions/${this.sessionId}/steps`,
      {
        method: "POST",
        headers: {
          "Content-Type": "application/json",
        },
        body: JSON.stringify(requestBody),
      }
    );

    const responseData = (await response.json()) as StepResponse;
    return {
      explanation: responseData.explanation,
      actions: responseData.actions,
    };
  }

  async retrieve(
    statement: string,
    accessibilityTree: string,
    title: string,
    url: string,
    screenshot?: string
  ): Promise<[string, Data]> {
    await this.ensureSession();
    const requestBody: StatementRequest = {
      statement,
      accessibility_tree: accessibilityTree,
      title,
      url,
      screenshot: screenshot || null,
    };
    const response = await this.fetchWithTimeout(
      `${this.baseUrl}/v1/sessions/${this.sessionId}/statements`,
      {
        method: "POST",
        headers: {
          "Content-Type": "application/json",
        },
        body: JSON.stringify(requestBody),
      }
    );

    const responseData = (await response.json()) as StatementResponse;
    return [responseData.explanation, looselyTypecast(responseData.result)];
  }

  async findArea(
    description: string,
    accessibilityTree: string
  ): Promise<{ id: number; explanation: string }> {
    await this.ensureSession();
    const requestBody: AreaRequest = {
      description,
      accessibility_tree: accessibilityTree,
    };
    const response = await this.fetchWithTimeout(
      `${this.baseUrl}/v1/sessions/${this.sessionId}/areas`,
      {
        method: "POST",
        headers: {
          "Content-Type": "application/json",
        },
        body: JSON.stringify(requestBody),
      }
    );

    const responseData = (await response.json()) as AreaResponse;
    return { id: responseData.id, explanation: responseData.explanation };
  }

  async findElement(
    description: string,
    accessibilityTree: string
  ): Promise<FindResponse["elements"][0]> {
    await this.ensureSession();
    const requestBody: FindRequest = {
      description,
      accessibility_tree: accessibilityTree,
    };
    const response = await this.fetchWithTimeout(
      `${this.baseUrl}/v1/sessions/${this.sessionId}/elements`,
      {
        method: "POST",
        headers: {
          "Content-Type": "application/json",
        },
        body: JSON.stringify(requestBody),
      }
    );

    const responseData = (await response.json()) as FindResponse;
    return responseData.elements[0];
  }

  async saveCache(): Promise<void> {
    await this.ensureSession();
    await this.fetchWithTimeout(
      `${this.baseUrl}/v1/sessions/${this.sessionId}/caches`,
      {
        method: "POST",
      }
    );
  }

  async discardCache(): Promise<void> {
    await this.ensureSession();
    await this.fetchWithTimeout(
      `${this.baseUrl}/v1/sessions/${this.sessionId}/caches`,
      {
        method: "DELETE",
      }
    );
  }

  async getStats(): Promise<StatsResponse> {
    await this.ensureSession();
    const response = await this.fetchWithTimeout(
      `${this.baseUrl}/v1/sessions/${this.sessionId}/stats`,
      {
        method: "GET",
      }
    );
    return (await response.json()) as StatsResponse;
  }
}
